lib/fluent/command/unpacker.rb

```diff
@@ -0,0 +1,253 @@
+
+See 'fluent-unpacker <command> --help' for more information on a specific command.
+HELP
+
+  def initialize(argv = ARGV)
+    @argv = argv
+  end
+
+  def call
+    command_class = Command.const_get(command)
+    command_class.new(@argv).call
+  end
+
+  private
+
+  def command
+    command = @argv.shift
+    if !command || !SUBCOMMAND.include?(command)
+      usage "Required subcommand : #{SUBCOMMAND.join(' | ')}"
+    end
+
+    command.split('_').map(&:capitalize).join('')
+  end
+
+  def usage(msg = nil)
+    puts HELP_TEXT
+    puts "Error: #{msg}" if msg
+    exit 1
+  end
+end
+
+module Command
+  class Base
+    def initialize(argv = ARGV)
+      @argv = argv
+
+      @options = {}
+      @opt_parser = OptionParser.new do |opt|
+        opt.separator 'Options:'
+
+        opt.on('-p DIR', '--plugins', 'add library path') do |v|
+          @options[:plugins] = v
+        end
+      end
+    end
+
+    def call
+      raise NotImplementedError, 'BUG: command  MUST implement this method'
+    end
+
+    private
+
+    def usage(msg = nil)
+      puts @opt_parser.to_s
+      puts "Error: #{msg}" if msg
+      exit 1
+    end
+
+    def parse_options!
+      ret = @opt_parser.parse(@argv)
+
+      if @options[:plugins] && !Dir.exist?(@options[:plugins])
+        usage "Directory #{@options[:plugins]} doesn't exist"
+      elsif @options[:plugins]
+        Fluent::Plugin.add_plugin_dir(@options[:plugins])
+      end
+
+      ret
+    end
+  end
+
+  module Formattable
+    DEFAULT_OPTIONS = {
+      format: :out_file
+    }
+
+    def initialize(argv = ARGV)
+      super
+
+      @options.merge!(DEFAULT_OPTIONS)
+      @params = {}
+      if i = @argv.index('--')
+        @params = @argv[i+1..-1].reduce({}) do |acc, e|
+          k, v = e[1..-1].split('=')
+          acc.merge(k => v)
+        end
+        @argv = @argv[0...i]
+      end
+
+      configure_option_parser
+    end
+
+    private
+
+    def configure_option_parser
+      @opt_parser.banner = "Usage: fluent-unpacker #{self.class.to_s.split('::').last.downcase} [options] file [-- <params>]"
+
+      @opt_parser.on('-f TYPE', '--format', 'configure output format') do |v|
+        @options[:format] = v.to_sym
+      end
+    end
+
+    def lookup_formatter(format, params)
+      conf = Fluent::Config::Element.new('ROOT', '', params, [])
+      formatter = Fluent::Plugin.new_formatter(format)
+
+      if formatter.respond_to?(:configure)
+        formatter.configure(conf)
+      end
+      formatter
+    rescue => e
+      usage e
+    end
+  end
+
+  class Head < Base
+    include Formattable
+
+    DEFAULT_HEAD_OPTIONS = {
+      count: 5
+    }
+
+    def initialize(argv = ARGV)
+      super
+      @options.merge!(DEFAULT_HEAD_OPTIONS)
+      @path = parse_options!
+    end
+
+    def call
+      @formatter = lookup_formatter(@options[:format], @params)
+
+      File.open(@path, 'r') do |io|
+        i = 0
+        MessagePack::Unpacker.new(io).each do |(time, record)|
```

Comment:
Don't use raw `MessagePack::Unpacker`, because we need to unpack values of ext type `Fluent::EventType`.
Use `Fluent::MessagePackFactory.unpacker` instead.